pennylane/dla/lie_closure.py

```diff
@@ -0,0 +1,291 @@
+            pw_to_idx (dict): map from ``PauliWord`` to row index in ``M``
+            rank (int): current rank of ``M``, equal to its number of columns
+            num_pw (int): current number of ``PauliWord``\ s, equal to the number of rows in ``M``
+        Returns:
+            ndarray: updated coefficient matrix for the LIS
+            dict: updated map from ``PauliWord`` to row index in ``M``. Includes new ``PauliWord`` keys
+                from the input ``pauli_sentence`` if it was linearly independent
+            int: updated rank/number of columns of ``M``
+            int: updated number of ``PauliWord``\ s/number of rows of ``M``
+            bool: whether ``pauli_sentence`` was linearly independent and its column was added to ``M``
+        """
+        new_pws = [pw for pw in pauli_sentence.keys() if pw not in pw_to_idx]
+        new_num_pw = num_pw + len(new_pws)
+
+        if new_num_pw < rank + 1:
+            # Can't span rank+1 independent vectors in fewer than rank+1 dimensions
+            # The input PauliSentence must have been linearly dependent
+            return M, pw_to_idx, rank, num_pw, False
+
+        M = np.pad(M, ((0, new_num_pw - num_pw), (0, 1)))
+
+        # If there are new PauliWords (i.e. new basis vectors), the candidate vector must be linearly independent
+        if new_num_pw > num_pw:
+            new_pw_to_idx = copy(pw_to_idx)
+            for i, pw in enumerate(new_pws, start=num_pw):
+                new_pw_to_idx[pw] = i
+            # Add new PauliSentence entries to matrix
+            for pw, value in pauli_sentence.items():
+                M[new_pw_to_idx[pw], rank] = value
+
+            return M, new_pw_to_idx, rank + 1, new_num_pw, True
+
+        # Add new PauliSentence entries to matrix
+        for pw, value in pauli_sentence.items():
+            M[pw_to_idx[pw], rank] = value
+
+        # Check if new vector is proportional to any of the previous vectors
+        # This is significantly cheaper than computing the rank and should be done first
+        if _is_any_col_propto_last(M):
+            M = M[:num_pw, :rank]
+            return M, pw_to_idx, rank, num_pw, False
+
+        new_rank = np.linalg.matrix_rank(M)  # expensive
+
+        # Manual singular value alternative, probably slower than ``matrix_rank``
+        # sing_value = np.min(np.abs(svd(M, compute_uv=False, lapack_driver="gesdd", check_finite=False)))
+        if new_rank == rank:
+            M = M[:num_pw, :rank]
+            return M, pw_to_idx, rank, num_pw, False
+
+        return M, pw_to_idx, rank + 1, new_num_pw, True
+
+    def __repr__(self):
+        return str(self.basis)
+
+    def __eq__(self, other):
+        """
+        Two VSpaces are equivalent when they span the same dimensional space.
+        This is checked here by having matching PauliWord keys in the sparse DOK representation and having the same rank.
```

Comment:
Just to understand: If the Pauli words match and the spanned dimension is equal, we consider the VSpace to be equal, even if the spanned subspace itself is not the same?

Reply:
Arent they the same in that case?
If M has the same rank and they overall concern the same pauli words, they should be equivalent no?
Like X(0), Y(0), Z(0) is equivalent to X(0) + Y(0), Y(0), Z(0)).
Can you give an example of what you mean?

Reply:
I was thinking of cases like `[X(0)+Y(0), Y(0) + Z(0)]` vs `[X(0)+Z(0), Y(0)]`, which both span 2D subspaces of the 3D space spanned by `[X(0), Y(0), Z(0)]` but the subspaces are not the same (I think?)

Reply:
Found this, looks good 👍 https://math.stackexchange.com/questions/1790121/how-to-check-if-two-sets-of-vectors-span-the-same-space-in-matlab-python will implement

Reply:
Added fix and test in [00afb95](https://github.com/PennyLaneAI/pennylane/pull/5161/commits/00afb95c5fd0a1d373be092bd4fe8f8f5cede4bc) and [ab5be97](https://github.com/PennyLaneAI/pennylane/pull/5161/commits/ab5be976210b4a78ff02b8c2070b12afbaf6a1ec)

Reply:
Very neat! :heart_eyes: